prefect_aws/filesystems.py

```diff
@@ -0,0 +1,201 @@
+        """
+
+        path = path or str(uuid4())
+
+        # If basepath provided, it means we won't write to the root dir of
+        # the bucket. So we need to add it on the front of the path.
+        path = str(Path(self.basepath) / path) if self.basepath else path
+
+        return path
+
+    def _get_s3_client(self) -> boto3.client:
+
+        """
+        Authenticate MinIO credentials or AWS credentials and return an S3 client.
+        This is a helper function called by read_path() or write_path().
+        """
+
+        s3_client_kwargs = {}
+
+        if self.minio_credentials:
+
+            aws_secret_access_key = self.minio_credentials.minio_root_password
+            s3_client_kwargs.update(
+                aws_access_key_id=self.minio_credentials.minio_root_user,
+                aws_secret_access_key=aws_secret_access_key.get_secret_value(),
+                endpoint_url=self.endpoint_url,
+            )
+
+        s3_client = boto3.client(service_name="s3", **s3_client_kwargs)
+
+        return s3_client
+
+    async def read_path(self, path: str) -> bytes:
+
+        """
+        Read specified path from S3 and return contents. Provide the entire
+        path to the key in S3.
+
+        Args:
+            - path: Entire path to (and including) the key.
+
+        Example:
+
+        If I have a bucket called "bucket" containing a folder called
+        "subfolder" which contains an object called "file1", I can
+        access its contents like so:
+
+        ```python
+        s3_bucket_block = S3Bucket(
+            bucket_name="bucket",
+            aws_credentials=AwsCredentials,
+            basepath="subfolder"
+        )
+
+        key_contents = s3_bucket_block.read_path(path="subfolder/file1")
+        ```
+        """
+
+        return await to_thread.run_sync(self._read_sync, path)
+
+    def _read_sync(self, key: str) -> bytes:
+
+        """
+        Called by read_path(). Creates an S3 client and retrieves the
+        contents from  a specified path.
+        """
+
+        s3_client = self._get_s3_client()
+
+        with io.BytesIO() as stream:
+
+            s3_client.download_fileobj(Bucket=self.bucket_name, Key=key, Fileobj=stream)
+            stream.seek(0)
+            output = stream.read()
+            return output
+
+    async def write_path(self, path: str, content: bytes) -> str:
+
+        """
+        Writes to an S3 bucket.
+
+        Args:
+
+        - path: The key name. Each object in your bucket has a unique
+        key (or key name).
+        - content: What you are uploading to S3.
+
+        Example:
+
+        ```python
+            s3_bucket_block = S3Bucket(
+                bucket_name="bucket",
+                minio_credentials=MinIOCredentials,
+                basepath="dogs/smalldogs",
+                endpoint_url="http://localhost:9000",
+            )
+            s3_havanese_path = s3_bucket_block.write_path(path="havanese", content=data)
+        ```
```

Comment:
```suggestion
            Write data to the path `dogs/small_dogs/havanese` in an S3 Bucket:
            ```python
                s3_bucket_block = S3Bucket(
                    bucket_name="bucket",
                    minio_credentials=MinIOCredentials,
                    basepath="dogs/smalldogs",
                    endpoint_url="http://localhost:9000",
                )
                s3_havanese_path = s3_bucket_block.write_path(path="havanese", content=data)
            ```
```